doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+- `database.OpenDB(dbstr)`: parses `dbstr` and returns the appropriate `DB` class.
+- `database.DB.Exec(self, stmt):` executes a query without returning any rows.
+- `database.DB.Query`(self, stmt): executes a query that returns `Rows` , typically a SELECT
+- `database.DB`: the object of a database connection
+- `database.Rows`: the object returned by `database.DB.Query`, should be a python generator
+
+##### Semantic Checking And Diagnostics
+
+As described in [The Flaws In The Current Architecture](#the-flaws-in-a-nutshell), SQLFlow executes ML tasks in python, but users are always only familiar with SQL. To generate user friendly diagnostics, we must check whether the user inputs satisfy the requirements of the python functions or callables that will be eventually called.
+
+The most widely-accepted solution to this problem is [contract programming](https://en.wikipedia.org/wiki/Design_by_contract). It enables defining formal, precise, and verifiable interface specifications for software components such as functions and callables. What we've done in the [sql/codegen/attribute](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/attribute) package is actually `contracts` in a limited way.
+
+###### Python
+
+We propose to design a contracts based mechanism in `contracts.py` for semantic checking and improving our daily development in python. The module should be able to:
+
+1. Check inputs for existing models thoroughly, for example, `canned estimators` or `xgboost`.
+2. Provide model developers an ease-to-use API to contribute models with thoroughly checking
+   - `docstring` or `decorator` 
+3. Throw an exception that incorporates all the violations to the contracts.
+4. Require a custom model to provide thoroughly checking for each input parameters.
+   - Making sure customized models will not reduce user experience
+
+Not all runtime errors can be catched by the contracts based design, for example, training binary classifiers with dataset that has three classes. Therefore, besides `contracts.py`, we need another mechanism to catch such errors and pass a user-friendly diagnostics messages to the user. The mechanism would be implemented in `diagnosics.py`.
+
+###### Golang
+
+The golang package `executor` should parse the diagnostics messages generated by `diagnostics.py` and return them to the users.
+
+In the remainder of the section, we say **diagnostics are required** where `diagnostics.py` should be involved to generate an error message.
+
+##### Intermediate Representations
+
+As described above, the proposed architecture requires the golang part to pass the *ir* of a SQLFlow program to the python part to execute the program. We propose to define `ir` as a protobuf file in `pkg/proto/ir.proto`. It should have at least the following messages and fields, each other detailed design doc could add necessary fields to the `Statement` message.
+
+```protobuf
+message Statement {
```

Comment:
It's simple to just add JSON/YAML hints in a Go struct so that we won't need to define a proto, the IR's struct in `ir.go` itself defines the JSON/YAML serialization constraints.

Reply:
I prefer protobuf because it can directly map to structs/classes in both python and golang. JSON/YAML is much more fragile if used as `dicts/lists` because they're less structural and lost the protection of the type system and the syntax of python.